tests/test_fork.py

```diff
@@ -0,0 +1,31 @@
+from pathlib import Path
+import fault
+import magma as m
+from .common import pytest_sim_params
+
+
+def pytest_generate_tests(metafunc):
+    pytest_sim_params(metafunc, 'system-verilog')
+
+
+def test_fork_join(target, simulator):
+    class dut(m.Circuit):
+        io = m.IO(I=m.In(m.Bits[2]), O=m.Out(m.Bits[2]))
+        io.O @= io.I
+
+    tester = fault.Tester(dut)
+    proc1 = tester.fork("proc1")
+    proc2 = tester.fork("proc2")
+
+    for i, proc in enumerate([proc1, proc2]):
+        proc.poke(tester.circuit.I[i], 1)
+        proc.eval()
+        proc.expect(tester.circuit.O[i], 1)
+
+    tester.join(proc1, proc2)
+
+    tester.compile_and_run(
+        target=target,
+        simulator=simulator,
+        tmp_dir=True
+    )
```

Comment:
LGTM, I wonder if we could add a test that requires the `fork/join` semantics to be correct for it to work?

Simple idea: Each thread sets an index (like this test does), master thread after join checks the entire output (all indices) are set.

Another idea: simple circuit that passes through ready to valid.  One thread sets ready after a random number of cycles, other thread loops until ready is high and sets valid, both join when ready/valid are high.

Reply:
I prefer the second idea. Is there a `wait` action in fault?

Reply:
There is a `wait_until_high` function (https://github.com/leonardt/fault/blob/master/fault/tester/staged_tester.py#L390) which I think does what you want? There's also `wait_on` which takes a generic condition.

Reply:
I tried `wait_on` and `wait_until_high` and both of them complains about clock not specified. SystemVerilog has a `wait` construct that blocks the current process execution until the signal is high, and that construct does not depends on clock, see LRM: 9.2. Here is a simple example of `wait`:

```SystemVerilog
module top;
logic a;
initial begin
    wait(a);
    $display("a is set @ %d", $time);
end
initial begin
    #10;
    a = 1;
end
endmodule
```

Reply:
Hmm, we can add support for `wait` in SV, I think that makes sense.

One problem though is how to map this concept to `verilator` which is a cycle accurate simulator? I guess one way to view it is `wait` will wait until the next simulator time step, so in verilator that would be every time `step` is called (normally this is defined w.r.t. to the clock in verilator, but in theory you could make `step` do something different if you were trying to model an analog circuit or something, but I'm not sure if this would work in the verilator model).  

Would the current `wait_on` implementation be equivalent to `wait` if we were looking at a purely synchronous design with one clock? Then one option is to stage `wait_on` in the frontend and compile based on the backend (so while loop in verilator, `wait` statement in system verilog)

Reply:
My understanding of Verilator is that `wait` semantics doesn't exists since the programming model assumes a single-thread driver code. After `eval()` is called, the state of circuit is always synchronized and you can just use `if` statement to test out inside a loop. Partially it is because Verilator is not an event-driven simulator, unlike commercial simulators. In my opinion, if there is no multi-threaded driver implementation in Verilator, we would never worry about `wait` semantics.